### PR TITLE
Trapsim update

### PR DESCRIPTION
**Pull request by bazinski**
- tracklet labels assigned and sent on, they are not usable though.
- time records sent with tracklets (have problems, hacked in sending code in subsequent pr )
- tracklet class modified trying to maintain backwards compatability and forward, not ideal.
- parallelisation removed, sorting after is too expensive, might be different for pbpb
- ignore raw data functions for now, they will come in a different PR

### CODE REVIEW

**Comment by bazinski** [comment]
Forgot labels are dodgy. Epiphanies at 2am...
The example I was running which has 1032 tracklets, 3329 digits and a mctruthcontainer index length of 701 for the tracklet labels, and 1120 labels. It does in fact *appear* to be working. Will lift the WIP and clean abit.

I will revert the changes the TRDDigitizer.cxx and do them/it seperately.

File: Detectors/TRD/base/include/TRDBase/Calibrations.h [hunk 1]
```diff
@@ -93,6 +94,11 @@ class Calibrations
   bool isChamberNotCalibrated(int det) const { return mChamberStatus->isNotCalibrated(det); }
   char getChamberStatusRaw(int det) const { return mChamberStatus->getStatus(det); }
 
+  //online gain tables.
+  const double getOnlineGainAdcdac(int det, int row, int mcm);
+  const double getOnlineGainFGAN(int det, int row, int mcm, int adc);
+  const double getOnlineGainFGFN(int det, int row, int mcm, int adc);
```
Comment (sawenzel) [inline Detectors/TRD/base/include/TRDBase/Calibrations.h:100]: const should be placed after the function

File: Detectors/TRD/base/include/TRDBase/PadParameters.h [hunk 2]
```diff
@@ -49,14 +50,14 @@ class PadParameters
   T getValue(int col, int row) const { return getValue(getChannel(col, row)); };
   void setValue(int ich, T value) { mData[ich] = value; }
   void setValue(int col, int row, T value) { setValue(getChannel(col, row), value); }
-  void debug() { std::cout << "in padparameters debug with this:" << this << " with mPlane:mChamber:mNrows:mNchannels::" << mPlane << ":" << mNrows << ":" << mNchannels << "   and a vector size of : " << mData.size() << std::endl; };
+  void debug() { LOG(debug) << "in padparameters debug with this:" << this << " with mPlane:mChamber:mNrows:mNchannels::" << mPlane << ":" << mNrows << ":" << mNchannels << "   and a vector size of : " << mData.size(); };
```
Comment (sawenzel) [inline Detectors/TRD/base/include/TRDBase/PadParameters.h:53]: do you need this in the header? best to avoid including logging header here.

File: Detectors/TRD/base/include/TRDBase/Tracklet.h [hunk 3]
```diff
@@ -40,48 +40,56 @@ namespace o2
...
   int getROB() const { return mROB; }
-  int getLabel() const { return mLabel[0]; }
-  int getLabel(const int i) const { return mLabel[i]; }
-  const std::array<int, 3>& getLabels() const { return mLabel; }
-  bool hasLabel(const int label) const { return (mLabel[0] == label || mLabel[1] == label || mLabel[2] == label); }
+  // labels out for now, index in digits taken out so cant reconcile.
+  //  int getLabel() const { return mLabel[0]; }
+  //  int getLabel(const int i) const { return mLabel[i]; }
+  //  const std::array<int, 3>& getLabels() const { return mLabel; }
+  //  bool hasLabel(const int label) const { return (mLabel[0] == label || mLabel[1] == label || mLabel[2] == label); }
```
Comment (sawenzel) [inline Detectors/TRD/base/include/TRDBase/Tracklet.h:92]: is this temporary? Why take out those functions in contrast to not using them?

File: Detectors/TRD/base/include/TRDBase/Tracklet.h [hunk 4]
```diff
@@ -101,8 +109,8 @@ class Tracklet
   int getNHits0() const { return mNHits0; }
   int getNHits1() const { return mNHits1; }
 
-  unsigned int getTrackletWord() const { return mTrackletWord; }
-  void setTrackletWord(unsigned int trackletWord) { mTrackletWord = trackletWord; }
+  unsigned int getTrackletWord() const;
+  void setTrackletWord(unsigned int trackletWord) { LOG(fatal) << "This is no longer permitted, o2::trd::Tracklet::setTrackletWord"; }
```
Comment (sawenzel) [inline Detectors/TRD/base/include/TRDBase/Tracklet.h:113]: why not remove this function. then nobody can call it.

File: Detectors/TRD/base/include/TRDBase/Tracklet.h [hunk 5]
```diff
@@ -114,6 +122,10 @@ class Tracklet
   void setNHits(int nhits) { mNHits = nhits; }
   void setNHits0(int nhits) { mNHits0 = nhits; }
   void setNHits1(int nhits) { mNHits1 = nhits; }
+  void setPID(float pid) { mPID = pid; }
+  void setZ(float z) { LOG(fatal) << "Cant call setZ in Tracklet class as its a property of the detector"; }
```
Comment (sawenzel) [inline Detectors/TRD/base/include/TRDBase/Tracklet.h:126]: why introducing the function and then putting a log(fatal)?

File: Detectors/TRD/base/macros/PrintRun2DigitsAndRun3 [hunk 6]
```diff
@@ -0,0 +1,112 @@
+#if !defined(__CINT__) || defined(__MAKECINT__)
+#include <TClonesArray.h>
```
Comment (sawenzel) [inline Detectors/TRD/base/macros/PrintRun2DigitsAndRun3:2]: Might it make sense to put these Run2 macros under a special folder?
Maybe it would still make sense to put the ".C" suffix and add these macros to the list of
non-checked macros.

File: Detectors/TRD/base/macros/PrintRun2DigitsAndRun3 [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+//#include "TRDDataFormat/TriggerRecord.h"
+#include "TRDBase/Digit.h"
+#include "TRDBase/Tracklet.h"
+
+using namespace o2;
+using namespace trd;
+using namespace std;
+
+
+void ConvertRun2DigitsAndTracklets()
```
Comment (sawenzel) [inline Detectors/TRD/base/macros/PrintRun2DigitsAndRun3:31]: this function name is the same as in another macro above. Is this intented?

File: Detectors/TRD/base/src/Tracklet.cxx [hunk 8]
```diff
@@ -26,74 +26,82 @@
...
-  if (mTrackletWord & 0x1000) {
-    return -((~(mTrackletWord - 1)) & 0x1fff);
-  } else {
-    return (mTrackletWord & 0x1fff);
-  }
+  //  if (mTrackletWord & 0x1000) {
+  //    return -((~(mTrackletWord - 1)) & 0x1fff);
+  //  } else {
+  //    return (mTrackletWord & 0x1fff);
+  //  }
```
Comment (sawenzel) [inline Detectors/TRD/base/src/Tracklet.cxx:85]: if this is no longer need I think it is better to delete comments and put this function in the header

File: Detectors/TRD/base/src/Tracklet.cxx [hunk 9]
```diff
@@ -26,74 +26,82 @@
...
-    return -((~((mTrackletWord >> 13) - 1)) & 0x7f);
-  } else {
-    return ((mTrackletWord >> 13) & 0x7f);
-  }
+  //  if (mTrackletWord & (1 << 19)) {
+  //    return -((~((mTrackletWord >> 13) - 1)) & 0x7f);
+  //  } else {
+  //    return ((mTrackletWord >> 13) & 0x7f);
+  //  }
+  return mdY;
```
Comment (sawenzel) [inline Detectors/TRD/base/src/Tracklet.cxx:97]: same as above

File: Detectors/TRD/simulation/src/TrapSimulator.cxx [hunk 10]
```diff
@@ -722,15 +741,23 @@ void TrapSimulator::draw(int choice, int index)
...
 
       trklLines[iTrkl].Draw();
     }
   }
+  //hist->Write();
+  // histfiltered->Write();
+  // for (int iTrkl = 0; iTrkl < mTrackletArray.size(); iTrkl++) trklLines[iTrkl].Write();
+  // grHits->Write();
+  //  hist->release();
+  //  grHits->release();
```
Comment (sawenzel) [inline Detectors/TRD/simulation/src/TrapSimulator.cxx:755]: do you need these new comments (which seem to be code)?

File: Detectors/TRD/trap/test_trap_tpp.cpp [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#include "trap_tpp.h"
+#include <stdio.h>
+#include <stdlib.h>
+
+// $Id$:
+
+int main(int argc, char** argv)
```
Comment (sawenzel) [inline Detectors/TRD/trap/test_trap_tpp.cpp:17]: what is this doing? It does not seem to be a unit test ... so maybe call it differently than "test_".

File: Detectors/TRD/trap/trap_tpp.cpp [hunk 12]
```diff
@@ -0,0 +1,575 @@
...
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#include "trap_tpp.h"
+#include <stdio.h>
+#include <math.h>
+
+// $Id$:
```
Comment (sawenzel) [inline Detectors/TRD/trap/trap_tpp.cpp:15]: Is this a new file? Consider stating where it comes from (if applicable).

File: Detectors/TRD/trap/trap_tpp.h [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#include <stdint.h>
```
Comment (sawenzel) [inline Detectors/TRD/trap/trap_tpp.h:11]: new file?

**Comment by sawenzel** [comment]
@tdietel : Could you please also take a look here?

File: Detectors/TRD/base/include/TRDBase/PadParameters.h (hunk 2 above)
Comment (bazinski) [reply Detectors/TRD/base/include/TRDBase/PadParameters.h:53]: no that was just a temp debugging thing late one night that slipped through, will remove.

File: Detectors/TRD/base/include/TRDBase/Tracklet.h (hunk 4 above)
Comment (bazinski) [reply Detectors/TRD/base/include/TRDBase/Tracklet.h:113]: cool will remove.

File: Detectors/TRD/base/macros/PrintRun2DigitsAndRun3 (hunk 6 above)
Comment (bazinski) [reply Detectors/TRD/base/macros/PrintRun2DigitsAndRun3:2]: Probably both suggestions make sense.

File: Detectors/TRD/trap/test_trap_tpp.cpp (hunk 11 above)
Comment (bazinski) [reply Detectors/TRD/trap/test_trap_tpp.cpp:17]: these are files for Venelin to have a stand alone c++ program for implementation of the assembler. I needed somewhere to put them so figured this is as good as any other.
A readme in the directory is probably called for. Else one up to explain the directory structure.

File: Detectors/TRD/trap/trap_tpp.cpp (hunk 12 above)
Comment (bazinski) [reply Detectors/TRD/trap/trap_tpp.cpp:15]: as comment directly above.

File: Detectors/TRD/trap/trap_tpp.h (hunk 13 above)
Comment (bazinski) [reply Detectors/TRD/trap/trap_tpp.h:11]: as comment directly above above.

File: Detectors/TRD/trap/test_trap_tpp.cpp (hunk 11 above)
Comment (bazinski) [reply Detectors/TRD/trap/test_trap_tpp.cpp:17]: Second thoughts will remove and put in a seperate PR

File: Detectors/TRD/base/include/TRDBase/Calibrations.h (hunk 1 above)
Comment (bazinski) [reply Detectors/TRD/base/include/TRDBase/Calibrations.h]: done

File: Detectors/TRD/base/include/TRDBase/PadParameters.h (hunk 2 above)
Comment (bazinski) [reply Detectors/TRD/base/include/TRDBase/PadParameters.h:53]: to remove the logging I need to create a source file and remove the functions at the bottom of the header file that have logging in them as well.

File: Detectors/TRD/base/include/TRDBase/Tracklet.h (hunk 5 above)
Comment (bazinski) [reply Detectors/TRD/base/include/TRDBase/Tracklet.h:126]: removed

File: Detectors/TRD/simulation/include/TRDSimulation/Digitizer.h [hunk 14]
```diff
@@ -38,6 +38,7 @@ struct SignalArray {
   double firstTBtime;                     // first TB time
   std::unordered_map<int, int> trackIds;  // tracks Ids associated to the signal
   std::vector<MCLabel> labels;            // labels associated to the signal
+  bool writtenOut{false};                 // whether this signal has been written out to an event.
```
Comment (jolopezl) [inline Detectors/TRD/simulation/include/TRDSimulation/Digitizer.h:41]: Please, remove this; all signals are written out, and the implementation of the pileup will take care of event triggering.

File: Detectors/TRD/base/include/TRDBase/Tracklet.h (hunk 4 above)
Comment (bazinski) [reply Detectors/TRD/base/include/TRDBase/Tracklet.h:113]: removed

File: Detectors/TRD/base/include/TRDBase/PadParameters.h (hunk 2 above)
Comment (bazinski) [reply Detectors/TRD/base/include/TRDBase/PadParameters.h:53]: removed

File: Detectors/TRD/simulation/src/Digitizer.cxx [hunk 15]
```diff
@@ -120,10 +120,10 @@ void Digitizer::flush(DigitContainer& digits, o2::dataformats::MCTruthContainer<
   for (int det = 0; det < kNdet; ++det) {
     std::move(mDigitsCollection[det].begin(), mDigitsCollection[det].end(), std::back_inserter(digits));
     for (const auto& it : mSignalsMapCollection[det]) {
-      std::move(it.second.labels.begin(), it.second.labels.end(), std::back_inserter(mMergedLabels));
+      labels.addElements(labels.getIndexedSize(), it.second.labels); 
     }
   }
-  labels.addElements(labels.getIndexedSize(), mMergedLabels);
+  //moved above to almost fix labels labels.addElements(labels.getIndexedSize(), mMergedLabels);
```
Comment (jolopezl) [inline Detectors/TRD/simulation/src/Digitizer.cxx:126]: Please, undo this changes, this doesn't fix the labels at all. I am fixing this in another PR.

File: Detectors/TRD/base/include/TRDBase/Calibrations.h (hunk 1 above)
Comment (bazinski) [reply Detectors/TRD/base/include/TRDBase/Calibrations.h]: fixed and the assocated implementations.

File: Detectors/TRD/workflow/src/TRDDigitizerSpec.cxx [hunk 16]
```diff
@@ -99,8 +99,11 @@ class TRDDPLDigitizerTask : public o2::base::BaseDPLDigitizer
         LOG(INFO) << "For collision " << collID << " eventID " << part.entryID << " found TRD " << hits.size() << " hits ";
 
         mDigitizer.process(hits, digits, labels);
+        mDigitizer.flush(digits, labels);
+        LOG(debug) << "For collision " << collID << " eventID " << part.entryID << " found TRD " << hits.size() << " hits with digits of size :" << digits.size();
```
Comment (jolopezl) [inline Detectors/TRD/workflow/src/TRDDigitizerSpec.cxx:103]: I would also remove all of this printing, we now what the problem is and how to fix it.

File: Detectors/TRD/workflow/src/TRDDigitReaderSpec.cxx [hunk 17]
```diff
@@ -89,10 +89,10 @@ void TRDDigitReaderSpec::run(ProcessingContext& pc)
     DPLTree->SetBranchAddress(mTriggerRecordBranchName.c_str(), &mPTriggerRecords);
     DPLTree->SetBranchAddress(mMCLabelsBranchName.c_str(), &mPMCLabels);
     DPLTree->GetEntry(0);
-    LOG(info) << "TRDDigitReader digits size=" << mDigits.size() << " triggerrecords size=" << mTriggerRecords.size() << " mc labels size=" << mMCLabels.getNElements();
     pc.outputs().snapshot(Output{"TRD", "DIGITS", 0, Lifetime::Timeframe}, mDigits);
     pc.outputs().snapshot(Output{"TRD", "TRGRDIG", 0, Lifetime::Timeframe}, mTriggerRecords);
     pc.outputs().snapshot(Output{"TRD", "LABELS", 0, Lifetime::Timeframe}, mMCLabels);
+    LOG(debug) << "TRDDigitReader digits size=" << mDigits.size() << " triggerrecords size=" << mTriggerRecords.size() << " mc labels size=" << mMCLabels.getNElements();
```
Comment (jolopezl) [inline Detectors/TRD/workflow/src/TRDDigitReaderSpec.cxx:95]: I would keep the INFO print level.

File: Detectors/TRD/base/src/Tracklet.cxx (hunk 8 above)
Comment (bazinski) [reply Detectors/TRD/base/src/Tracklet.cxx]: removed

File: Detectors/TRD/base/src/Tracklet.cxx (hunk 9 above)
Comment (bazinski) [reply Detectors/TRD/base/src/Tracklet.cxx:86]: removed

File: Detectors/TRD/base/include/TRDBase/Tracklet.h (hunk 3 above)
Comment (bazinski) [reply Detectors/TRD/base/include/TRDBase/Tracklet.h]: labels moved to an external MCTruthContainer, although it might seem more logical to keep them with their respective tracklets.

File: Detectors/TRD/simulation/src/TrapSimulator.cxx (hunk 10 above)
Comment (bazinski) [reply Detectors/TRD/simulation/src/TrapSimulator.cxx]: gone

File: Detectors/TRD/simulation/include/TRDSimulation/Digitizer.h (hunk 14 above)
Comment (bazinski) [reply Detectors/TRD/simulation/include/TRDSimulation/Digitizer.h:41]: oops that was me buggering around trying to fix labels.
removing.

File: Detectors/TRD/simulation/src/Digitizer.cxx (hunk 15 above)
Comment (bazinski) [reply Detectors/TRD/simulation/src/Digitizer.cxx:126]: it fixes it in so far as i dont get *all* the labels on the first digit, which messes up the trap simulator. this was discussed last thursday.

File: Detectors/TRD/workflow/src/TRDDigitizerSpec.cxx (hunk 16 above)
Comment (bazinski) [reply Detectors/TRD/workflow/src/TRDDigitizerSpec.cxx:103]: removed

File: Detectors/TRD/workflow/src/TRDDigitReaderSpec.cxx (hunk 17 above)
Comment (bazinski) [reply Detectors/TRD/workflow/src/TRDDigitReaderSpec.cxx:95]: done

File: Detectors/TRD/simulation/src/Digitizer.cxx (hunk 15 above)
Comment (bazinski) [reply Detectors/TRD/simulation/src/Digitizer.cxx:126]: hang on this was removed from this pr? let me check, to go in later.

Comment (bazinski) [reply Detectors/TRD/simulation/src/Digitizer.cxx]: out, it was reverted already in a previous push, not sure how it crept back in.

File: Detectors/TRD/base/macros/PrintRun2DigitsAndRun3 (hunk 6 above)
Comment (bazinski) [reply Detectors/TRD/base/macros/PrintRun2DigitsAndRun3:2]: That entire directory is aliroot dependent, will make it clear in a readme.

File: Detectors/TRD/base/include/TRDBase/PadParameters.h (hunk 2 above)
Comment (bazinski) [reply Detectors/TRD/base/include/TRDBase/PadParameters.h]: removed

File: Detectors/TRD/base/macros/PrintRun2DigitsAndRun3 (hunk 7 above)
Comment (bazinski) [reply Detectors/TRD/base/macros/PrintRun2DigitsAndRun3]: this one should actually not be there and is now removed, and rested added to the ExclusionList

File: Detectors/TRD/workflow/src/TRDDigitizerSpec.cxx [hunk 18]
```diff
@@ -99,8 +99,11 @@ class TRDDPLDigitizerTask : public o2::base::BaseDPLDigitizer
         LOG(INFO) << "For collision " << collID << " eventID " << part.entryID << " found TRD " << hits.size() << " hits ";
 
         mDigitizer.process(hits, digits, labels);
+        mDigitizer.flush(digits, labels);
```
Comment (jolopezl) [inline Detectors/TRD/workflow/src/TRDDigitizerSpec.cxx:102]: You can't force flushing the digits here; this will break embedding.

File: Detectors/TRD/workflow/src/TRDDigitizerSpec.cxx [hunk 19]
```diff
@@ -111,14 +114,16 @@ class TRDDPLDigitizerTask : public o2::base::BaseDPLDigitizer
         labels.clear();
       }
     }
-
+    LOG(info) << " digits : " << digits.size() << " labels labelssize : " << labels.getNElements() << " labels index size : " << labels.getIndexedSize();
     // Force flush of the digits that remain in the digitizer cache
     mDigitizer.flush(digits, labels);
-    triggers.emplace_back(irecords[irecords.size() - 1], digitsAccum.size(), digits.size());
+    LOG(info) << " digits : " << digits.size() << " labels labelssize : " << labels.getNElements() << " labels index size : " << labels.getIndexedSize();
+    //triggers.emplace_back(irecords[irecords.size() - 1], digitsAccum.size(), digits.size());
```
Comment (jolopezl) [inline Detectors/TRD/workflow/src/TRDDigitizerSpec.cxx:121]: This part of the code requieres a cleanup, way too much printing left after debuting, maybe.

File: Detectors/TRD/workflow/src/TRDDigitizerSpec.cxx (hunk 18 above)
Comment (bazinski) [reply Detectors/TRD/workflow/src/TRDDigitizerSpec.cxx:102]: ja that is a carry over from me buggering around

Comment (bazinski) [reply Detectors/TRD/workflow/src/TRDDigitizerSpec.cxx]: I thought I reverted this back to the other one late last night

File: Detectors/TRD/workflow/src/TRDDigitizerSpec.cxx (hunk 19 above)
Comment (bazinski) [reply Detectors/TRD/workflow/src/TRDDigitizerSpec.cxx]: its reverted to dev branch again.

**Comment by bazinski** [comment]
@sawenzel will make a comprehensive commit message when its all finalised. will rebase and collapse the above messages.

**Comment by tdietel** [comment]
I have noticed that the PR includes the old online gain tables from run 1+2. Do you plan to keep them in there, or are they mainly for testing based on the old data?

**Comment by tdietel** [comment]
I think we will need to clarify the class naming for o2::trd::Tracklet. As pointed out in the header file, this is mostly for debugging, and it would be good to reflect that in the name.

We will get a conflict when we define the tracklet format that goes into the CTF.

But maybe that's for a later PR.

File: Detectors/TRD/base/src/FeeParam.cxx [hunk 20]
```diff
@@ -252,7 +252,7 @@ int FeeParam::getMCMfromPad(int irow, int icol) const
 
   if (irow < 0 || icol < 0 || irow > mgkNrowC1 || icol > mgkNcol)
     return -1;
-
+  //LOG(info) << "getMCMfromPad row : " << irow << " col : "<< icol << "answer : "<< (icol % (mgkNcol / 2)) / mgkNcolMcm + mgkNmcmRobInCol * (irow % mgkNmcmRobInRow)<< " with : " << (icol % (144 / 2)) << "/"<< 18 <<"+"<< 4 <<"*"<< (irow % mgkNmcmRobInRow);
```
Comment (tdietel) [inline Detectors/TRD/base/src/FeeParam.cxx:255]: This should be removed

File: Detectors/TRD/base/src/FeeParam.cxx [hunk 21]
```diff
@@ -292,8 +292,8 @@ int FeeParam::getROBfromPad(int irow, int icol) const
   //
   // Return on which rob this pad is
   //
-
-  return (irow / mgkNmcmRobInRow) * 2 + getColSide(icol);
+  //LOG(info) << "getROBfromPad  row : "<< irow << "col : " << icol << " answer : " <<  (int)((int)irow / (int)mgkNmcmRobInRow) * 2 + getColSide(icol)<< " with : " << (irow / mgkNmcmRobInRow )*2 << " + " << getColSide(icol);
```
Comment (tdietel) [inline Detectors/TRD/base/src/FeeParam.cxx:295]: This should be removed

File: Detectors/TRD/simulation/include/TRDSimulation/TrapConfig.h [hunk 22]
```diff
@@ -34,7 +35,7 @@ namespace trd
 class TrapConfig
 {
  public:
-  TrapConfig(std::string configname);
+  TrapConfig(std::string configname = "default");
```
Comment (tdietel) [inline Detectors/TRD/simulation/include/TRDSimulation/TrapConfig.h:38]: What is the "default" config? It would be better to provide the actual name.

**Comment by tdietel** [comment]
> @sawenzel will make a comprehensive commit message when its all finalised. will rebase and collapse the above messages.

I agree with Sandro, this is a huge PR that involves several items that are rather loosely connected. We should try to have more smaller, but more frequent PRs.

File: Detectors/TRD/base/src/FeeParam.cxx (hunk 20 above)
Comment (bazinski) [reply Detectors/TRD/base/src/FeeParam.cxx:255]: removed and changed some others from info to debug as they should be.

File: Detectors/TRD/base/src/FeeParam.cxx (hunk 21 above)
Comment (bazinski) [reply Detectors/TRD/base/src/FeeParam.cxx:295]: removed as per comment previous

Comment (bazinski) [reply Detectors/TRD/base/src/FeeParam.cxx:295]: they should be debug bug are commented out for performance reasons, apparently my sed did not find those.

File: Detectors/TRD/simulation/include/TRDSimulation/TrapConfig.h (hunk 22 above)
Comment (bazinski) [reply Detectors/TRD/simulation/include/TRDSimulation/TrapConfig.h:38]: Its a carry over from the old way of initialising the trapconfig via the trapconfighandler which is an open question whether we need trapconfighandler. The default is specified on the command line to the trapsimulator dpl device. Removing.

**Comment by bazinski** [comment]
> I think we will need to clarify the class naming for o2::trd::Tracklet. As pointed out in the header file, this is mostly for debugging, and it would be good to reflect that in the name.
> 
> We will get a conflict when we define the tracklet format that goes into the CTF.
> 
> But maybe that's for a later PR.
Good point. I had intended the structure for tracklets to be o2::trd::RawTracklet, but the complication comes in that some of the tracklet data sits in the header o2::trd::RawTrapHeader.
Ja another PR in the wings.

**Comment by bazinski** [comment]
@tdietel 
> I have noticed that the PR includes the old online gain tables from run 1+2. Do you plan to keep them in there, or are they mainly for testing based on the old data?
Backwards testing, and yes, also it does not hurt to have it for when someone decides they want to revert to do it again.

**Comment by bazinski** [comment]
> > @sawenzel will make a comprehensive commit message when its all finalised. will rebase and collapse the above messages.
> 
> I agree with Sandro, this is a huge PR that involves several items that are rather loosely connected. We should try to have more smaller, but more frequent PRs.

The only loosely connected stuff was Venelin stuff which crept into this PR.
The rest of the stuff is interdependent. I cant change the tracklet with out changing its dependent classes.

The primary original reason for this PR is the labels and time being sent with the tracklets, the rest is all dependencies.
So this fixes 3 things
labels
triggerrecords
Trapconfigs.

The temp fix to incoming labels and trigger records are out so ironically none of this makes it workable for @martenole. Trigger records are fixed in next pr inside the trapsimulator.

File: Detectors/TRD/simulation/include/TRDSimulation/TrapConfig.h (hunk 22 above)
Comment (bazinski) [reply Detectors/TRD/simulation/include/TRDSimulation/TrapConfig.h]: trapconfighandler code is still there, just not used anymore in device.

**Comment by tdietel** [comment]
> The only loosely connected stuff was Venelin stuff which crept into this PR.
> The rest of the stuff is interdependent. I cant change the tracklet with out changing its dependent classes.

Ok, sounds good.
 
> The primary original reason for this PR is the labels and time being sent with the tracklets, the rest is all dependencies.
> So this fixes 3 things
> labels
> triggerrecords
> Trapconfigs.

Ok, and this has been tested on your private version of the digitizer and is expected to work.

> The temp fix to incoming labels and trigger records are out so ironically none of this makes it workable for @martenole. Trigger records are fixed in next pr inside the trapsimulator.

Hopefully the PR by @jolopezl will fix this.

**Comment by sawenzel** [comment]
So we go on here?

**Comment by sawenzel** [comment]
Let's go on in order to unblock the other 2 PRs for TRD.
